### PR TITLE
add reward manager

### PR DESCRIPTION
**Pull request by yhnsu**
# Description

This document summarizes the usage of **RewardManager** and **ObservationManager** in our RL training pipeline, following a simple and practical format for contributors.

---

## Summary of Change

- Clarifies how RewardManager and ObservationManager are integrated into RL training.
- Provides motivation/context for modular reward and observation handling.
- No new dependencies required.

---

## Motivation & Context

- **RewardManager** and **ObservationManager** modularize reward and observation logic for RL tasks.
- They allow flexible configuration via JSON, making it easy to add, remove, or tune reward/observation terms without changing code.
- This design supports rapid experimentation and reproducibility in RL research.

---

### RewardManager
- Reads reward configuration from the environment config (e.g., `gym_config.json`).
- Each reward term is defined by a function, weight, and parameters.
- During each RL step, RewardManager computes all active reward terms and combines them (weighted sum or replace mode).
- Individual reward components are logged for analysis (e.g., via wandb/tensorboard).
- Example config:
  ```json
  "rewards": {
    "distance_reward": {
      "func": "distance_to_target",
      "mode": "add",
      "weight": 0.5,
      "params": {"source_entity_cfg": {"uid": "cube"}, "target_pose_key": "goal_pose"}
    },
    "success_bonus": {
      "func": "success_reward",
      "mode": "add",
      "weight": 10.0,
      "params": {}
    }
  }
  ```
- RL trainer logs both total reward and each component for debugging and analysis.

### ObservationManager
- Reads observation configuration from the environment config.
- Each observation term is defined by a function and parameters.
- During each RL step, ObservationManager collects and processes all active observation terms, producing the final observation dict.
- Observations are flattened for RL algorithms and logged if needed.
- Example config:
  ```json
  "observations": {
    "cube_pos": {
      "func": "get_cube_position",
      "mode": "add",
      "params": {"entity_cfg": {"uid": "cube"}}
    },
    "robot_state": {
      "func": "get_robot_state",
      "mode": "add",
      "params": {"entity_cfg": {"uid": "Manipulator"}}
    }
  }
  ```
- RL trainer uses the flattened observation for policy and buffer.

---

### CODE REVIEW

File: embodichain/lab/gym/envs/managers/observations.py [hunk 1]
```diff
@@ -252,6 +252,52 @@ def compute_semantic_mask(
     return torch.stack(masks, dim=-1)
 
 
+def get_robot_ee_pose(
```
Comment (yuecideng) [inline embodichain/lab/gym/envs/managers/observations.py]: Use `get_robot_eef_pose`

File: embodichain/lab/gym/envs/managers/rewards.py [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+def reward_from_obs(
+    env: EmbodiedEnv,
+    obs: dict,
+    action: torch.Tensor,
+    info: dict,
+    obs_key: str = "robot/qpos",
+    target_value: float = 0.0,
+    scale: float = 1.0,
+) -> torch.Tensor:
+    """Reward based on observation values."""
```
Comment (yuecideng) [inline embodichain/lab/gym/envs/managers/rewards.py]: Add docs for this function

File: embodichain/lab/gym/envs/managers/rewards.py [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+import torch
+from typing import TYPE_CHECKING
+
+from embodichain.lab.gym.envs.managers.cfg import SceneEntityCfg
+
+if TYPE_CHECKING:
+    from embodichain.lab.gym.envs import EmbodiedEnv
+
+
+def reward_from_obs(
```
Comment (yuecideng) [inline embodichain/lab/gym/envs/managers/rewards.py]: Can not understand the meanning of this reward founction

File: embodichain/lab/gym/envs/managers/rewards.py [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+    env: EmbodiedEnv,
+    obs: dict,
+    action: torch.Tensor,
+    info: dict,
+    source_entity_cfg: SceneEntityCfg = None,
+    target_entity_cfg: SceneEntityCfg = None,
+    exponential: bool = False,
+    sigma: float = 1.0,
+) -> torch.Tensor:
+    """Reward based on distance between two entities."""
```
Comment (yuecideng) [inline embodichain/lab/gym/envs/managers/rewards.py]: Add full docs for this function

File: embodichain/lab/gym/envs/managers/rewards.py [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+
+def joint_velocity_penalty(
+    env: EmbodiedEnv,
+    obs: dict,
+    action: torch.Tensor,
+    info: dict,
+    robot_uid: str = "robot",
+    joint_ids: slice | list[int] = slice(None),
+) -> torch.Tensor:
+    """Penalize large joint velocities."""
```
Comment (yuecideng) [inline embodichain/lab/gym/envs/managers/rewards.py]: Add full docs for this function

File: configs/agents/rl/push_cube/gym_config.json [hunk 6]
```diff
@@ -13,20 +13,103 @@
...
+        "params": {
+          "target_pose_key": "goal_pose"
+        }
+      }
+    },
+    "rewards": {
+      "reaching_reward": {
+        "func": "reaching_behind_object_reward",
+        "mode": "add",
+        "name": "reaching",
```
Comment (yuecideng) [inline configs/agents/rl/push_cube/gym_config.json]: I think the arg `name` is duplicated, since we already have this reward functor key

File: embodichain/lab/gym/envs/managers/rewards.py (hunk 3 above)
Comment (yhnsu) [reply embodichain/lab/gym/envs/managers/rewards.py]: already removed

File: configs/agents/rl/push_cube/gym_config.json (hunk 6 above)
Comment (yhnsu) [reply configs/agents/rl/push_cube/gym_config.json]: already removed `name`
